modules/client/src/main/java/org/apache/ignite/internal/client/table/ClientPartitionManager.java

```diff
@@ -0,0 +1,138 @@
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.apache.ignite.internal.client.proto.ClientMessageUnpacker;
+import org.apache.ignite.internal.client.proto.ClientOp;
+import org.apache.ignite.internal.table.partition.HashPartition;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.table.Tuple;
+import org.apache.ignite.table.mapper.Mapper;
+import org.apache.ignite.table.partition.Partition;
+import org.apache.ignite.table.partition.PartitionManager;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Client partition manager implementation.
+ */
+public class ClientPartitionManager implements PartitionManager {
+    private final ClientTable tbl;
+
+    private final Lock lock = new ReentrantLock();
+
+    private final Map<Partition, ClusterNode> cache = new HashMap<>();
+
+    private @Nullable Instant aliveUntil;
+
+    ClientPartitionManager(ClientTable clientTable) {
+        this.tbl = clientTable;
+    }
+
+    @Override
+    public CompletableFuture<ClusterNode> primaryReplicaAsync(Partition partition) {
+        if (!(partition instanceof HashPartition)) {
+            throw new IllegalArgumentException("Unsupported partition type: " + partition);
+        }
+
+        ClusterNode clusterNode = getClusterNode(partition);
+
+        if (clusterNode != null) {
+            return completedFuture(clusterNode);
+        }
+
+        return primaryReplicasAsync().thenApply(map -> {
+            lock.lock();
+            try {
+                cache.putAll(map);
+                aliveUntil = Instant.now().plus(1, ChronoUnit.MINUTES);
+                return map.get(partition);
+            } finally {
+                lock.unlock();
+            }
+        });
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    public CompletableFuture<Map<Partition, ClusterNode>> primaryReplicasAsync() {
+        return tbl.channel().serviceAsync(ClientOp.PRIMARY_REPLICAS_GET,
+                w -> w.out().packInt(tbl.tableId()),
+                r -> {
+                    ClientMessageUnpacker in = r.in();
+
+                    if (in.tryUnpackNil()) {
+                        return null;
+                    }
+
+                    int size = in.unpackInt();
+
+                    Map<Partition, ClusterNode> res = new HashMap<>(size);
+
+                    for (int i = 0; i < size; i++) {
+                        int partition = in.unpackInt();
+                        res.put(new HashPartition(partition), unpackClusterNode(r));
+                    }
+
+                    return res;
+                });
+    }
+
+    @Override
+    public <K> CompletableFuture<Partition> partitionAsync(K key, Mapper<K> mapper) {
```

Comment:
1. `key` can't be null, let's add validation.
2. `mapper` can be null (right?), let's add `@Nullable`.